imap_processing/ultra/l2/ultra_l2.py

```
@@ -148,12 +171,114 @@ def get_variable_attributes_optional_energy_dependence(
+    Parameters
+    ----------
+    pset : xarray.Dataset
+        Ultra L1C pointing set dataset to bin.
+    bin_groups : numpy.ndarray, optional
+        Array of indices defining the new energy bin edges. If not provided,
+        DEFAULT_BIN_GROUPS will be used.
+
+    Returns
+    -------
+    xarray.Dataset
+        The input pset with energy bins grouped according to the bin_groups.
+    numpy.ndarray
+        The new energy bin edges.
+    """
+    if bin_groups is None:
+        bin_groups = DEFAULT_BIN_GROUPS
+    # Get a list of variables that have the energy bin dimension
+    energy_dep_vars = [
+        var for var in pset.data_vars if "energy_bin_geometric_mean" in pset[var].dims
+    ]
+    # From those, get the variables that need to be averaged over the coarse bins
+    vars_to_average = [
+        var
+        for var in energy_dep_vars
+        if var in VARIABLES_TO_AVERAGE_OVER_COARSE_ENERGY_BINS
+    ]
+    n_fine_bins = pset["energy_bin_geometric_mean"].size
+    logger.info(
+        f"Binning pset fine energy bins into coarser bins with edges: {bin_groups}"
+    )
+
+    if bin_groups[-1] > n_fine_bins:
+        raise ValueError(
+            "The given bin_groups contain an index larger than the number of "
+            f"fine energy bins in the pset: {n_fine_bins}."
+        )
+    # Add any missing 'catch-all' bins at the start or end
+    if bin_groups[-1] < n_fine_bins:
+        bin_groups = np.append(bin_groups, n_fine_bins)
+        logger.info(
+            "The given bin_groups did not include the last fine energy bins."
+            f" Adding another 'catch-all' bin. New coarse energy bin edges:"
+            f" {bin_groups}"
+        )
+    if bin_groups[0] > 0:
+        bin_groups = np.insert(bin_groups, 0, 0)
+        logger.info(
+            "The given bin_groups did not include the first fine energy bins."
+            f" Adding another 'catch-all' bin. New coarse energy bin edges:"
+            f" {bin_groups}"
+        )
```

Aren't you doing all of this in the earlier utility function as well? I'm wondering if this is all necessary, or if you can raise and tell the user to provide the correct values instead of you guessing and doing a catch-all.

---

Which util? I asked Nick about this senario and he mentioned that they would prefer a catch-all bin.

---

I was thinking the `build_energy_bins()` function, but now that I look at it that is maybe just for the psets and not L2? But then above this you use: `build_default_coarse_bin_edges()`

If someone explicitly passes in the bin edges, then I don't think we should be adding the low/high into it for them, that was their perogative to ignore the low/high energies. Maybe someone wants a map with only 1 energy step from the mapper tool.

Basically, I think this is handled by your "default energy bins" function isn't it? And otherwise someone is explicit in what they asked for and we should honor that explicitness.

---

OK I see your point 👍